pkg/ring/ring.go

```diff
@@ -326,16 +326,37 @@ func (r *Ring) GetAll(op Operation) (ReplicationSet, error) {
 		return ReplicationSet{}, ErrEmptyRing
 	}
 
+	var maxUnavailableZones int
+	var numRequired int
+	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	if r.cfg.ZoneAwarenessEnabled {
+		zones := make(map[string]bool)
+
+		for _, ingester := range r.ringDesc.Ingesters {
+			zones[ingester.Zone] = true
+			ingesters = append(ingesters, ingester)
+		}
+		maxUnavailableZones = len(zones) / 2
+		// Assume ingesters are evenly spread over zones
+		// The maximum number of failures is all the ingesters in the maximum number of failing zones
+		maxErrors := (len(ingesters) / len(zones)) * maxUnavailableZones
+
+		return ReplicationSet{
```

Comment:
Design-wise I'm not a big fan of this early return. I would be glad if you could investigate if we can cleanly improve the `GetAll()` implementation without falling into two different implementations (in this PR it's basically a big if branch). If it's not clean, then we could get rediscuss it.

Reply:
The zoneware code was moved to be in the main code so it reuses more of the variables.